### PR TITLE
ddl: set correct startHandle when add indices meets some errors

### PR DESCRIPTION
**Pull request by winkyao**
Thank you for working on TiDB! Please read TiDB's [CONTRIBUTING](https://github.com/pingcap/tidb/blob/master/CONTRIBUTING.md) document **BEFORE** filing this PR.

## What have you changed? (mandatory)

Before this PR, we will fail the added gofail test.

When we encounter an error, we will update incorrect handle to the DDL job. Then we will never backfill the skipped indices range, cause `admin check table` failed.

## What are the type of the changes (mandatory)?

- Bug fix (non-breaking change which fixes an issue)

## How has this PR been tested (mandatory)?

Unit test.

@zimulala @crazycs520 @ciscoxll PTAL

### CODE REVIEW

File: ddl/index.go [hunk 1]
```diff
@@ -815,32 +832,32 @@ func (w *worker) handleReorgTasks(startTime time.Time, startHandle int64, reorgI
...
 			endIncluded = true
 		}
-		task := &reorgIndexTask{startHandle, endHandle, endIncluded}
+		task := &reorgIndexTask{startHandle, endHandle, endIncluded, false}
 
 		batchTasks = append(batchTasks, task)
 		if len(batchTasks) >= len(workers) || i == (len(kvRanges)-1) {
 			// Wait tasks finish.
-			err = w.handleReorgTasks(startTime, startHandle, reorgInfo, &totalAddedCount, workers, batchTasks)
+			err = w.handleReorgTasks(startTime, firstHandle, reorgInfo, &totalAddedCount, workers, batchTasks)
```
Comment (tiancaiamao) [inline ddl/index.go]: I guess the `firstHandle` argument is redundant.
You can always get it from `batchTasks[0].start`, right?

File: ddl/index.go [hunk 2]
```diff
@@ -687,6 +689,14 @@ func (w *addIndexWorker) run(d *ddlCtx) {
 		}
 
 		log.Debug("[ddl-reorg] got backfill index task:#v", task)
+
+		if task.mockTask {
```
Comment (tiancaiamao) [inline ddl/index.go]: Why not gofail here directly?

Comment (winkyao) [reply ddl/index.go]: Here is in a multi-thread environment, so it's non-trivial to control the gofail run only once.

**Comment by winkyao** [comment]
@tiancaiamao PTAL

**Comment by tiancaiamao** [comment]
LGTM

File: ddl/index.go [hunk 3]
```diff
@@ -779,14 +789,23 @@ func (w *worker) waitTaskResults(workers []*addIndexWorker, taskCnt int, totalAd
...
 
+var gofailMockAddindexErrOnceGuard bool
+
 // handleReorgTasks sends tasks to workers, and waits for all the running workers to return results,
 // there are taskCnt running workers.
-func (w *worker) handleReorgTasks(startTime time.Time, startHandle int64, reorgInfo *reorgInfo, totalAddedCount *int64, workers []*addIndexWorker, batchTasks []*reorgIndexTask) error {
+func (w *worker) handleReorgTasks(reorgInfo *reorgInfo, totalAddedCount *int64, workers []*addIndexWorker, batchTasks []*reorgIndexTask) error {
 	for i, task := range batchTasks {
+		// gofail: var mockAddIndexErr bool
+		// if mockAddIndexErr && !gofailMockAddindexErrOnceGuard {
```
Comment (shenli) [inline ddl/index.go]: `gofailMockAddindexErrOnceGuard` is too long.

File: ddl/index.go (hunk 2 above)
Comment (coocood) [reply ddl/index.go]: @winkyao 
I think it worth it to gofail here directly.
We can check the worker id to avoid race.

File: ddl/index.go (hunk 3 above)
Comment (winkyao) [reply ddl/index.go]: To avoid name conflicts.

**Comment by winkyao** [comment]
@zimulala @coocood PTAL

**Comment by coocood** [comment]
LGTM

**Comment by crazycs520** [comment]
LGTM

**Pull request merged**